### ISSUE TITLE
Deleting item changes child item icon

### ISSUE THREAD

**Issue by stdedos** [shundhammer/qdirstat#114]
While testing #113, I saw this:

Deleting e.g. `AutoSave` dir, changes treeview from this:
![image](https://user-images.githubusercontent.com/133706/61632985-cdd6f780-ac96-11e9-9aa3-690404a2d5a2.png)

to that:
![image](https://user-images.githubusercontent.com/133706/61633034-e6dfa880-ac96-11e9-8dfd-218f0d37e77a.png)

_(i.e. Plugins folder is now a "drive" rather than a folder)_

Logfile:
```


2019-07-22 15:37:10.526 [1965] <Info>    Logger.cpp:143 openLogFile():  -- Log Start --
2019-07-22 15:37:10.526 [1965] <Info>    main.cpp:58 logVersion():  QDirStat-1.5.92-git built with Qt 5.5.1
2019-07-22 15:37:10.702 [1965] <WARNING> [Qt] [31mvoid DBusMenuExporterPrivate::addAction(QAction*, int)[0m: Already tracking action "" under id 46
2019-07-22 15:37:10.706 [1965] <Debug>   DebugHelpers.cpp:133 dumpExcludeRules():  <ExcludeRule ".snapshot">
2019-07-22 15:37:10.707 [1965] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by NameCol ascending
2019-07-22 15:37:10.707 [1965] <Debug>   TreemapView.cpp:123 setSelectionModel():  
2019-07-22 15:37:10.708 [1965] <Info>    Cleanup.cpp:415 desktopSpecificApps():  Detected desktop "Unity"
2019-07-22 15:37:10.708 [1965] <Info>    Cleanup.cpp:468 desktopSpecificApps():  %filemanager => "nautilus"
2019-07-22 15:37:10.708 [1965] <Info>    Cleanup.cpp:468 desktopSpecificApps():  %terminal => "gnome-terminal &"
2019-07-22 15:37:10.708 [1965] <Debug>   MainWindow.cpp:1224 changeLayout():  Changing to layout L2
2019-07-22 15:37:10.710 [1965] <Info>    PkgQuery.cpp:66 checkPkgManagers():  Checking available supported package managers...
2019-07-22 15:37:10.710 [1965] <Debug>   SysUtil.cpp:97 runCommand():  /usr/bin/dpkg -S /usr/bin/dpkg
2019-07-22 15:37:10.884 [1965] <Info>    PkgQuery.cpp:95 checkPkgManager():  Found primary package manager dpkg
2019-07-22 15:37:10.884 [1965] <Info>    SysUtil.cpp:83 runCommand():  Command not found: /bin/rpm
2019-07-22 15:37:10.884 [1965] <Info>    SysUtil.cpp:83 runCommand():  Command not found: /usr/bin/pacman
2019-07-22 15:37:10.884 [1965] <Info>    PkgQuery.cpp:84 checkPkgManagers():  Found dpkg
2019-07-22 15:37:10.884 [1965] <Info>    MainWindow.cpp:1282 toggleVerboseSelection():  Verbose selection is now off. Change this with Shift-F7.
2019-07-22 15:37:12.470 [1965] <Debug>   DirTree.cpp:480 setExcludeRules():  Clearing tmp exclude rules
2019-07-22 15:37:12.470 [1965] <Info>    DirTree.cpp:121 startReading():     url: "/home/sntentos"
2019-07-22 15:37:12.471 [1965] <Info>    DirTree.cpp:124 startReading():  device: /home/sntentos/.Private
2019-07-22 15:37:12.471 [1965] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-22 15:37:12.525 [1965] <Debug>   DirReadJob.cpp:241 startReading():  Found cache file .qdirstat.cache.gz
2019-07-22 15:37:12.525 [1965] <Debug>   DirReadJob.cpp:391 readCacheFile():  Using cache file /home/sntentos/.qdirstat.cache.gz for /home/sntentos
2019-07-22 15:37:12.525 [1965] <Debug>   DirReadJob.cpp:398 readCacheFile():  Clearing complete tree
2019-07-22 15:37:12.679 [1965] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-22 15:37:12.731 [1965] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-22 15:37:16.704 [1965] <Debug>   DirTreeCache.cpp:257 ~CacheReader():  Cache reading finished
2019-07-22 15:37:16.956 [1965] <Info>    MainWindow.cpp:641 readingFinished():  
2019-07-22 15:37:16.981 [1965] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by PercentNumCol descending
2019-07-22 15:37:16.981 [1965] <Debug>   MainWindow.cpp:623 idleDisplay():  No current branch - expanding tree to level 1
2019-07-22 15:37:16.981 [1965] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-22 15:37:17.076 [1965] <Info>    MainWindow.cpp:647 readingFinished():  Reading finished after  4.465 sec
2019-07-22 15:39:23.103 [1965] <Debug>   OutputWindow.cpp:40 OutputWindow():  Creating
2019-07-22 15:39:23.112 [1965] <Info>    Trash.cpp:195 trash():  Successfully moved to trash: /home/sntentos/.audacity-data/AutoSave
2019-07-22 15:39:23.115 [1965] <Debug>   OutputWindow.cpp:321 noMoreProcesses():  Emitting lastProcessFinished() err: 0
2019-07-22 15:39:23.115 [1965] <Debug>   Refresher.cpp:39 refresh():  Refreshing 1 items
2019-07-22 15:39:23.115 [1965] <Debug>   FileInfoSet.cpp:60 invalidRemoved():  Keeping /home/sntentos/.audacity-data
2019-07-22 15:39:23.115 [1965] <Debug>   DirTreeModel.cpp:1041 clearingSubtree():  Deleting all children of /home/sntentos/.audacity-data
2019-07-22 15:39:23.115 [1965] <Debug>   DirTreeModel.cpp:1050 clearingSubtree():  beginRemoveRows for /home/sntentos/.audacity-data row 0 to 2
2019-07-22 15:39:23.115 [1965] <Debug>   DirTreeModel.cpp:1081 invalidatePersistent():  Invalidating <QModelIndex row: 0 col: 0 /home/sntentos/.audacity-data/AutoSave >
2019-07-22 15:39:23.141 [1965] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by NameCol ascending
2019-07-22 15:39:23.141 [1965] <Debug>   OutputWindow.cpp:308 closeIfDone():  No more processes to watch. Auto-closing.
2019-07-22 15:39:23.202 [1965] <Debug>   OutputWindow.cpp:57 ~OutputWindow():  Destructor
2019-07-22 15:39:23.206 [1965] <Info>    DirReadJob.cpp:113 crossingFileSystems():  File system boundary at mount point /home/sntentos/.audacity-data/Plug-Ins on device <unknown>
2019-07-22 15:39:23.242 [1965] <Info>    MainWindow.cpp:641 readingFinished():  
2019-07-22 15:39:23.248 [1965] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by PercentNumCol descending
2019-07-22 15:39:23.299 [1965] <Info>    MainWindow.cpp:647 readingFinished():  Reading finished after  0.136 sec
2019-07-22 15:40:59.528 [1965] <Info>    Logger.cpp:84 ~Logger():  -- Log End --

```

**Comment by shundhammer** [comment]
That directory was detected as a file system boundary, so it got the "mount point" icon:

```
2019-07-22 15:39:23.206 [1965] <Info>    DirReadJob.cpp:113 crossingFileSystems():  File system boundary at mount point /home/sntentos/.audacity-data/Plug-Ins on device <unknown>
```

That might legally happen since a mount point is not marked as such in a cache file.
Or it might be the problem that we saw before that on your system it does not seem to detect the devices of mounted filesystem correctly (thus the `unknown device` log entries).
That might have to do with the number of indirections with LVM and encryption.

**Comment by stdedos** [comment]
I could try submitting more debug information, if there is something specific that would help (not the tree itself though, apologies)

Otherwise, feel free to handle it as you wish (known issue, not-a-bug, invalid etc)

**Comment by shundhammer** [comment]
Well, is it a mount point or not? If it is, then this is just a minor discrepancy between information from a cache file and the real thing from the live filesystem.

If it's not a mount point, then I guess its needs some more handling for fringe cases where the device name cannot be determined (i.e. the "unknown device" case).

Is the "continue reading at mount point" action (context menu) enabled for that directory? Does it continue descending into that subtree when you use it? And if so, does it stop at the next directory level already?

**Comment by shundhammer** [comment]
The `df` command can tell you what device it thinks the directory is on. Please try that for that directory, for its parent and, if there are any, for the next subdirectory level (one or two subdirectories will be enough):

```bash
df -h -T /home/sntentos/.audacity-data
df -h -T /home/sntentos/.audacity-data/Plug-Ins
df -h -T /home/sntentos/.audacity-data/Plug-Ins/*
```

**Comment by stdedos** [comment]
> Well, is it a mount point or not? If it is, then this is just a minor discrepancy between information from a cache file and the real thing from the live filesystem.

`/home` is a LVM mount point, `~` is an `ecryptfs` mount point too. The rest, AFAIK, are not.

I'll try and add all the requested diagnostics/information tomorrow.

**Comment by shundhammer** [comment]
One big difference from the normal use case is that you are starting from values from a QDirStat cache file, and then re-reading that directory branch from disk after the cleanup action reads the real values from the real disk.

The cache file, however, does not contain the device minor/major numbers (defaulting to 0) whereas the `stat()` syscall returns the real one in the `st_dev` field of the `struct stat` that it returns, so the check if this might be a mount point is getting more detailed.

https://github.com/shundhammer/qdirstat/blob/master/src/DirReadJob.cpp#L97

That check then tries to figure out what device both directories are on. For the parent, it traverses the tree upward to find the nearest mount point; for the child, it uses the child's path directly.

And this is where the trouble starts: Since the child is not really a mount point, there is no entry for it in the list of known mount points. We only got into that branch of the program because the initial test failed for completely the wrong reasons: The device numbers mismatched, but not because the directories reside on different devices, but because for the parent, the device number is completely unknown because the value was read from a cache file.

**Comment by shundhammer** [comment]
Now checking if the parent was a cached directory if the child device name is empty.

Please test.

**Comment by shundhammer** [comment]
Looks like you are the first user who fearlessly uses the cache files with all features that QDirStat has to offer. :smiley: So some problems are to be expected. Others seem to use it only to analyze, but not to clean up.

BTW if you like that feature so much, have you tried using the `qdirstat-cache-writer` script to create the cache files? You can do that in a cron job over night so you can automate that job.

See also

https://github.com/shundhammer/qdirstat/tree/master/scripts

**Comment by stdedos** [comment]
> Looks like you are the first user who fearlessly uses the cache files with all features that QDirStat has to offer. 😃

Well ... sort-of. It's the first time I am ever investigating that filesystem (or any non-Windows filesystem) to that extend, and detail; thus I treat it as my playground. I also mentioned it takes a bit of scanning time, and I wanted to keep restarting qdirstat, so ... it sounded logical.

I am also using the program as if it was TreeSize (https://www.jam-software.com/treesize_free/), since I am only familiar (and happy) with that. I learn the broad strokes of anything by using, not by reading man pages [1].

> So some problems are to be expected. Others seem to use it only to analyze, but not to clean up.

I understand that. But, you seem welcoming in fixing / creating features / "defining" behavior, so, why not?

I am also not using anything damaging (`rm`), except on throwaway things, so ... I should be okay. Theoretically. (Also, Jesus saves, but God makes backups, so ... I should be okay. Theoretically).

-----

> BTW if you like that feature so much, have you tried using the `qdirstat-cache-writer` script to create the cache files? You can do that in a cron job over night so you can automate that job.

Well .. I simply reported the bugs I've seen during "normal usage" ([^1]). I guess, when program "is perfect", or, I don't have any strong wishes over some functionality (or I completely clean up my system?), then I'll stop 😄 

-----

> See also
> 
> [/scripts@`master`](https://github.com/shundhammer/qdirstat/tree/master/scripts) [](/shundhammer/qdirstat/tree/HEAD@{2019-07-22T18:42:37Z}/scripts)

Seeing these files, I wanted to focus on https://github.com/shundhammer/qdirstat/blob/master/doc/QDirStat-for-Servers.md#L77-L80

I guess both things should be fixed sometime / somehow. At least, there should be a crude way to determine this

> they'd still run on your desktop machine instead of your server

If the cache file is not named as default `.qdirstat.cache.gz`, and the top-level cache directory does not match the enclosing directory, then [a] print a warning message and/or a question (Is this file from a remote server [Y/n]?) and, if so, [b] block all "removal" operations (or all operations probably? Except `AssumeRemoved`, I don't see any other valid operation for a remote cache file)

> There is also no indication that your are seeing the contents of a cache file

This also sounds trivial. Add an extra "info" label, permanently visible on the UI?

(I can split these on new or existing tickets, if you'd plan to do something about them)

**Comment by stdedos** [comment]
> ```shell
> df -h -T /home/sntentos/.audacity-data
> df -h -T /home/sntentos/.audacity-data/Plug-Ins
> df -h -T /home/sntentos/.audacity-data/Plug-Ins/
> ```

It's all the same:
```bash
$ mkdir -p a/b
$ df -h -T /home/sntentos/.audacity-data
Filesystem              Type      Size  Used Avail Use% Mounted on
/home/sntentos/.Private ecryptfs  709G  457G  217G  68% /home/sntentos
$ df -h -T /home/sntentos/.audacity-data/Plug-Ins
Filesystem              Type      Size  Used Avail Use% Mounted on
$ df -h -T /home/sntentos/.audacity-data/Plug-Ins/*
Filesystem              Type      Size  Used Avail Use% Mounted on
/home/sntentos/.Private ecryptfs  709G  457G  217G  68% /home/sntentos
$ df -h -T /home/sntentos/.audacity-data/Plug-Ins/a/
Filesystem              Type      Size  Used Avail Use% Mounted on
/home/sntentos/.Private ecryptfs  709G  457G  217G  68% /home/sntentos
$ df -h -T /home/sntentos/.audacity-data/Plug-Ins/a/b
Filesystem              Type      Size  Used Avail Use% Mounted on
/home/sntentos/.Private ecryptfs  709G  457G  217G  68% /home/sntentos
```

> Is the "continue reading at mount point" action (context menu) enabled for that directory? Does it continue descending into that subtree when you use it? And if so, does it stop at the next directory level already?

It only exists on "cached load". Regardless if `Plug-Ins` have a deeper structure (i.e. `Plug-Ins/a/b`) or not, it appears as a mounted point. Clicing "continue reading at mount point", it further loads the directory structure (and still has `Plug-Ins` marked as a mount point)

```


2019-07-23 13:06:30.789 [22681] <Info>    Logger.cpp:143 openLogFile():  -- Log Start --
2019-07-23 13:06:30.789 [22681] <Info>    main.cpp:58 logVersion():  QDirStat-1.6.01-git built with Qt 5.5.1
2019-07-23 13:06:30.925 [22681] <WARNING> [Qt] [31mvoid DBusMenuExporterPrivate::addAction(QAction*, int)[0m: Already tracking action "" under id 46
2019-07-23 13:06:30.929 [22681] <Debug>   DebugHelpers.cpp:133 dumpExcludeRules():  <ExcludeRule ".snapshot">
2019-07-23 13:06:30.929 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by NameCol ascending
2019-07-23 13:06:30.930 [22681] <Debug>   TreemapView.cpp:123 setSelectionModel():  
2019-07-23 13:06:30.930 [22681] <Info>    Cleanup.cpp:415 desktopSpecificApps():  Detected desktop "Unity"
2019-07-23 13:06:30.930 [22681] <Info>    Cleanup.cpp:468 desktopSpecificApps():  %filemanager => "nautilus"
2019-07-23 13:06:30.930 [22681] <Info>    Cleanup.cpp:468 desktopSpecificApps():  %terminal => "gnome-terminal &"
2019-07-23 13:06:30.930 [22681] <Debug>   MainWindow.cpp:1224 changeLayout():  Changing to layout L2
2019-07-23 13:06:30.932 [22681] <Info>    PkgQuery.cpp:66 checkPkgManagers():  Checking available supported package managers...
2019-07-23 13:06:30.933 [22681] <Debug>   SysUtil.cpp:97 runCommand():  /usr/bin/dpkg -S /usr/bin/dpkg
2019-07-23 13:06:31.110 [22681] <Info>    PkgQuery.cpp:95 checkPkgManager():  Found primary package manager dpkg
2019-07-23 13:06:31.110 [22681] <Info>    SysUtil.cpp:83 runCommand():  Command not found: /bin/rpm
2019-07-23 13:06:31.110 [22681] <Info>    SysUtil.cpp:83 runCommand():  Command not found: /usr/bin/pacman
2019-07-23 13:06:31.110 [22681] <Info>    PkgQuery.cpp:84 checkPkgManagers():  Found dpkg
2019-07-23 13:06:31.110 [22681] <Info>    MainWindow.cpp:1282 toggleVerboseSelection():  Verbose selection is now off. Change this with Shift-F7.
2019-07-23 13:06:34.140 [22681] <Debug>   DirTree.cpp:480 setExcludeRules():  Clearing tmp exclude rules
2019-07-23 13:06:34.140 [22681] <Info>    DirTree.cpp:121 startReading():     url: "/home/sntentos"
2019-07-23 13:06:34.141 [22681] <Info>    DirTree.cpp:124 startReading():  device: /home/sntentos/.Private
2019-07-23 13:06:34.141 [22681] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-23 13:06:34.186 [22681] <Debug>   DirReadJob.cpp:254 startReading():  Found cache file .qdirstat.cache.gz
2019-07-23 13:06:34.186 [22681] <Debug>   DirReadJob.cpp:404 readCacheFile():  Using cache file /home/sntentos/.qdirstat.cache.gz for /home/sntentos
2019-07-23 13:06:34.186 [22681] <Debug>   DirReadJob.cpp:411 readCacheFile():  Clearing complete tree
2019-07-23 13:06:34.358 [22681] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-23 13:06:34.393 [22681] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-23 13:06:38.208 [22681] <Debug>   DirTreeCache.cpp:257 ~CacheReader():  Cache reading finished
2019-07-23 13:06:38.438 [22681] <Info>    MainWindow.cpp:641 readingFinished():  
2019-07-23 13:06:38.460 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by PercentNumCol descending
2019-07-23 13:06:38.460 [22681] <Debug>   MainWindow.cpp:623 idleDisplay():  No current branch - expanding tree to level 1
2019-07-23 13:06:38.460 [22681] <Debug>   MainWindow.cpp:940 expandTreeToLevel():  Expanding tree to level 1
2019-07-23 13:06:38.493 [22681] <Info>    MainWindow.cpp:647 readingFinished():  Reading finished after  4.279 sec
2019-07-23 13:10:13.781 [22681] <ERROR>   ActionManager.cpp:55 action():  No action with name actionCopyUrlToClipboard found
2019-07-23 13:10:16.749 [22681] <Debug>   OutputWindow.cpp:40 OutputWindow():  Creating
2019-07-23 13:10:16.762 [22681] <Info>    Trash.cpp:195 trash():  Successfully moved to trash: /home/sntentos/.audacity-data/AutoSave
2019-07-23 13:10:16.766 [22681] <Debug>   OutputWindow.cpp:321 noMoreProcesses():  Emitting lastProcessFinished() err: 0
2019-07-23 13:10:16.766 [22681] <Debug>   Refresher.cpp:39 refresh():  Refreshing 1 items
2019-07-23 13:10:16.766 [22681] <Debug>   FileInfoSet.cpp:60 invalidRemoved():  Keeping /home/sntentos/.audacity-data
2019-07-23 13:10:16.766 [22681] <Debug>   DirTreeModel.cpp:1041 clearingSubtree():  Deleting all children of /home/sntentos/.audacity-data
2019-07-23 13:10:16.766 [22681] <Debug>   DirTreeModel.cpp:1050 clearingSubtree():  beginRemoveRows for /home/sntentos/.audacity-data row 0 to 2
2019-07-23 13:10:16.766 [22681] <Debug>   DirTreeModel.cpp:1081 invalidatePersistent():  Invalidating <QModelIndex row: 2 col: 0 /home/sntentos/.audacity-data/<Files> >
2019-07-23 13:10:16.766 [22681] <Debug>   DirTreeModel.cpp:1081 invalidatePersistent():  Invalidating <QModelIndex row: 1 col: 0 /home/sntentos/.audacity-data/AutoSave >
2019-07-23 13:10:16.797 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by NameCol ascending
2019-07-23 13:10:16.797 [22681] <Debug>   OutputWindow.cpp:308 closeIfDone():  No more processes to watch. Auto-closing.
2019-07-23 13:10:16.834 [22681] <Debug>   OutputWindow.cpp:57 ~OutputWindow():  Destructor
2019-07-23 13:10:16.840 [22681] <Info>    DirReadJob.cpp:126 crossingFileSystems():  File system boundary at mount point /home/sntentos/.audacity-data/Plug-Ins on device <unknown>
2019-07-23 13:10:16.870 [22681] <Info>    MainWindow.cpp:641 readingFinished():  
2019-07-23 13:10:16.876 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by PercentNumCol descending
2019-07-23 13:10:16.937 [22681] <Info>    MainWindow.cpp:647 readingFinished():  Reading finished after  0.112 sec
2019-07-23 13:10:24.728 [22681] <ERROR>   ActionManager.cpp:55 action():  No action with name actionCopyUrlToClipboard found
2019-07-23 13:10:26.178 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by NameCol ascending
2019-07-23 13:10:26.178 [22681] <Debug>   DirTreeModel.cpp:1110 refreshSelected():  Refreshing /home/sntentos/.audacity-data/Plug-Ins
2019-07-23 13:10:26.322 [22681] <Info>    MainWindow.cpp:641 readingFinished():  
2019-07-23 13:10:26.329 [22681] <Debug>   DirTreeModel.cpp:588 sort():  Sorting by PercentNumCol descending
2019-07-23 13:10:26.372 [22681] <Info>    MainWindow.cpp:647 readingFinished():  Reading finished after  0.153 sec
2019-07-23 13:10:42.357 [22681] <Info>    Logger.cpp:84 ~Logger():  -- Log End --

```

**Comment by shundhammer** [comment]
This should now be more reliable with the fix for #273.
